rust/hyperlane-base/src/types/multisig.rs

```diff
@@ -171,51 +203,218 @@ impl MultisigCheckpointSyncer {
                     }
 
                     // Ensure that the signature is actually by the validator
-                    let signer = signed_checkpoint.recover()?;
+                    let signer =
+                        self.fetch_or_recover(&signed_checkpoint, &mut recovered_addresses)?;
 
-                    if H256::from(signer) != *validator {
+                    if signer != vw.validator {
                         debug!(
-                            validator = format!("{:#x}", validator),
+                            validator = format!("{:#x}", vw.validator),
                             index = index,
                             "Checkpoint signature mismatch"
                         );
                         continue;
                     }
 
+                    // let checkpoint_hash = signed_checkpoint.value.signing_hash();
+
                     // Push the signed checkpoint into the hashmap
                     let root = signed_checkpoint.value.root;
                     let signed_checkpoints = signed_checkpoints_per_root.entry(root).or_default();
                     signed_checkpoints.push(signed_checkpoint);
 
+                    let cumulative_weight: Weight = signed_checkpoints
+                        .iter()
+                        .filter_map(|sc| {
+                            let signer =
+                                self.fetch_or_recover(sc, &mut recovered_addresses).ok()?;
```

Comment:
I should've been more clear in the last review but I'd prefer to not have any recovery for stuff like this, even if it's cached. I think it's more confusing than adding some additional data to `signed_checkpoints`